src/providers/aws.rs

```diff
@@ -0,0 +1,887 @@
+                            instance_id: Some(instance_id),
+                            public_dns_name: Some(public_dns),
+                            public_ip_address: Some(public_ip),
+                            ..
+                        } => {
+                            trace!(log, "instance ready";
+                                "instance_id" => instance_id.clone(),
+                                "ip" => &public_ip,
+                            );
+
+                            let (ipinfo, _) = self.instances.get_mut(&instance_id).unwrap();
+                            *ipinfo = Some((public_ip.clone(), public_dns.clone()));
+                        }
+                        _ => {
+                            all_ready = false;
+                        }
+                    }
+                }
+            }
+
+            if let Some(to) = max_wait {
+                if time::Instant::now().duration_since(start) > to {
+                    bail!("timed out");
+                }
+            }
+        }
+
+        use rayon::prelude::*;
+        self.instances
+            .par_iter()
+            .try_for_each(|(_instance_id, (ipinfo, (name, m_setup)))| {
+                let (public_ip, _) = ipinfo.as_ref().unwrap();
+                if let Setup { setup: Some(f), .. } = m_setup {
+                    super::setup_machine(
+                        log,
+                        &name,
+                        &public_ip,
+                        "ubuntu",
+                        max_wait,
+                        Some(private_key_path.path()),
+                        f.as_ref(),
+                    )?;
+                }
+
+                Ok(())
+            })
+    }
+
+    /// Establish SSH connections to the machines. The `Ok` value is a `HashMap` associating the
+    /// friendly name for each `Setup` with the corresponding SSH connection.
+    pub fn connect_all<'l>(&'l self) -> Result<HashMap<String, crate::Machine<'l>>, Error> {
+        let log = self.log.as_ref().unwrap();
+        let private_key_path = self.private_key_path.as_ref().unwrap();
+        self.instances
+            .values()
+            .map(|info| match info {
+                (Some((public_ip, public_dns)), (name, Setup { .. })) => {
+                    let mut m = Machine {
+                        public_ip: public_ip.clone(),
+                        public_dns: public_dns.clone(),
+                        nickname: name.clone(),
+                        ssh: None,
+                        _tsunami: Default::default(),
+                    };
+
+                    m.connect_ssh(log, "ubuntu", Some(private_key_path.path()))?;
+                    Ok((name.clone(), m))
+                }
+                _ => bail!("Machines not initialized"),
+            })
+            .collect()
+    }
+}
+
+impl std::ops::Drop for RegionLauncher {
+    fn drop(&mut self) {
```

Comment:
We should probably describe this dropping behavior somewhere.

Reply:
https://github.com/jonhoo/tsunami/pull/22/commits/9e6f09a2ef9f013e171272b892d80b6ba12bc97b#diff-60699702750aef5a8cdc050b4fce64eaR286